### PR TITLE
HAAR-821: added '/groups/{group}' delete group endpoint

### PR DESCRIPTION
**Pull request by ravmoj**
added '/groups/{group}' delete group endpoint

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/GroupsService.kt [hunk 1]
```diff
@@ -90,19 +95,51 @@ class GroupsService(
...
+          "GroupDeleteSuccess",
+          mapOf("username" to authenticationFacade.currentUsername, "groupCode" to groupCode),
+          null
+        )
+      }
+      else -> {
+        throw GroupHasChildGroupException(groupCode, "child group exist")
+      }
+    }
+  }
```
Comment (AlTurner-MOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/GroupsService.kt:125]: Better if all exceptions are not nested?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupService.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      Hibernate.initialize(u.groups)
+      u.groups.forEach { Hibernate.initialize(it.children) }
+      u.groups
+    }.orElse(null)
+  }
+
+  fun getAssignableGroups(username: String?, authorities: Collection<GrantedAuthority>): List<Group> =
+    if (canMaintainAuthUsers(authorities)) allGroups.toList()
+    else getAuthUserGroups(username)?.sortedBy { it.groupName } ?: listOf()
+
```
Comment (AlTurner-MOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupService.kt:84]: Do these exceptions need to be nested? Also why open?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/model/UserFilter.kt [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      andBuilder.add(groupPredicate)
+    }
+    if (!name.isNullOrBlank()) {
+      andBuilder.add(buildNamePredicate(root, cb))
+    }
+    if (status != Status.ALL) {
+      andBuilder.add(cb.equal(root.get<Any>("enabled"), status == Status.ACTIVE))
+    }
+    query.distinct(true)
+//    val personJoin = root.join<Any, Any>("person", JoinType.INNER)
```
[the remark this answers is not in the document]
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/model/UserFilter.kt:50]: Removed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/security/MaintainUserCheck.kt [hunk 4]
```diff
@@ -8,7 +9,13 @@ import uk.gov.justice.digital.hmpps.externalusersapi.jpa.repository.UserReposito
 class MaintainUserCheck(
   private val userRepository: UserRepository,
 ) {
-  @Throws(AuthGroupRelationshipException::class)
+
+  companion object {
+    fun canMaintainAuthUsers(authorities: Collection<GrantedAuthority>): Boolean =
```
[the remark this answers is not in the document]
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/security/MaintainUserCheck.kt:14]: renamed to canMaintainUsers

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/GroupsControllerIntTest.kt [hunk 5]
```diff
@@ -418,4 +418,62 @@ class GroupsControllerIntTest : IntegrationTestBase() {
...
+              "errorCode" to null,
+              "moreInfo" to null,
+              "userMessage" to "Unable to delete group: GC_DEL_3 with reason: child group exist",
+              "developerMessage" to "Unable to delete group: GC_DEL_3 with reason: child group exist"
+            )
+          )
+        }
+    }
+    @Test
+    fun `Delete Child Group endpoint returns forbidden when dose not have admin role`() {
```
[the remark this answers is not in the document]
Comment (ravmoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/GroupsControllerIntTest.kt:460]: changed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/GroupsService.kt (hunk 1 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/GroupsService.kt:125]: agreed ..will move it out as our standard

**Review by AlTurner-MOJ — APPROVED** [review]
lgtm

**Pull request merged**